File: qr/native/android/app/src/main/java/com/example/benchmark/qr/nativeapp/QrImages.kt
package com.example.benchmark.qr.nativeapp

import android.content.Context
import java.io.File

/** What the image at [index] encodes; see bench/scripts/make-qr-images.swift. */
fun expectedPayload(index: Int): String =
  "https://example.com/benchmark/qr/%03d".format(index)

/**
 * Copies the PNGs bundled under the `qr` assets directory into the app's own
 * storage, and returns the paths in name order.
 *
 * The benchmark decodes from file paths (see AGENTS.md), and an APK asset has
 * no path of its own. This is the preparation step, which is not measured.
 */
fun unpackImages(context: Context): List<String> {
  val destination = File(context.filesDir, "qr")
  destination.mkdirs()
  val names = (context.assets.list("qr") ?: emptyArray()).filter { it.endsWith(".png") }.sorted()
  return names.map { name ->
    val file = File(destination, name)
    // An empty file means a half-finished copy from an earlier launch.
    if (!file.exists() || file.length() == 0L) {
      context.assets.open("qr/$name").use { input ->
        file.outputStream().use { output -> input.copyTo(output) }
      }
    }
    file.absolutePath
  }
}

File: qr/native/ios/QrApp/QrImages.swift
import Foundation

/// What the image at `index` encodes; see bench/scripts/make-qr-images.swift.
func expectedPayload(_ index: Int) -> String {
  String(format: "https://example.com/benchmark/qr/%03d", index)
}

/// Copies the bundled images into Documents and returns the paths in name
/// order.
///
/// Every implementation decodes from a file path (see AGENTS.md). The images
/// are already files here, but they are unpacked anyway so all four start from
/// the same place. This is the preparation step, which is not measured.
func unpackImages() throws -> [String] {
  guard let source = Bundle.main.resourceURL?.appendingPathComponent("qr") else {
    throw QrImagesError.missingBundle
  }
  let documents = try FileManager.default.url(
    for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
  let destination = documents.appendingPathComponent("qr")
  try FileManager.default.createDirectory(at: destination, withIntermediateDirectories: true)

  let names = try FileManager.default.contentsOfDirectory(atPath: source.path)
    .filter { $0.hasSuffix(".png") }
    .sorted()
  return try names.map { name in
    let file = destination.appendingPathComponent(name)
    // An empty file means a half-finished copy from an earlier launch.
    let size = (try? file.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
    if size == 0 {
      try? FileManager.default.removeItem(at: file)
      try FileManager.default.copyItem(at: source.appendingPathComponent(name), to: file)
    }
    return file.path
  }
}

enum QrImagesError: LocalizedError {
  case missingBundle

  var errorDescription: String? {
    switch self {
    case .missingBundle: return "bundled images are missing"
    }
  }
}
